File: apps/software-architect-roadmap/examples/catalog/java/Catalog.java
import com.sun.net.httpserver.HttpServer;
import java.net.InetSocketAddress;
import java.nio.charset.StandardCharsets;

public class Catalog {
    public static void main(String[] args) throws Exception {
        String host = System.getenv().getOrDefault("BIND_HOST", "127.0.0.1");
        int port = Integer.parseInt(System.getenv().getOrDefault("PORT", "8081"));
        HttpServer server = HttpServer.create(new InetSocketAddress(host, port), 0);
        server.createContext("/", exchange -> {
            String path = exchange.getRequestURI().getPath();
            String body;
            int status;
            if (!exchange.getRequestMethod().equals("GET")) {
                body = "{\"error\":\"method not allowed\"}"; status = 405;
            } else if (path.equals("/products")) {
                body = "[{\"sku\":\"book\",\"unit_price\":1200},{\"sku\":\"pen\",\"unit_price\":200}]"; status = 200;
            } else if (path.equals("/health")) {
                body = "{\"status\":\"ok\"}"; status = 200;
            } else { body = "{\"error\":\"not found\"}"; status = 404; }
            byte[] bytes = body.getBytes(StandardCharsets.UTF_8);
            exchange.getResponseHeaders().set("Content-Type", "application/json");
            exchange.sendResponseHeaders(status, bytes.length);
            try (var stream = exchange.getResponseBody()) { stream.write(bytes); }
        });
        server.start();
        System.out.println("Catalog listening on " + host + ":" + server.getAddress().getPort());
    }
}

File: apps/software-architect-roadmap/examples/catalog/python/catalog.py
import json
import os
from http.server import BaseHTTPRequestHandler, ThreadingHTTPServer

PRODUCTS = [{"sku": "book", "unit_price": 1200}, {"sku": "pen", "unit_price": 200}]

class Handler(BaseHTTPRequestHandler):
    def send(self, status, body):
        raw = json.dumps(body).encode()
        self.send_response(status)
        self.send_header('Content-Type', 'application/json')
        self.send_header('Content-Length', str(len(raw)))
        self.end_headers()
        self.wfile.write(raw)
    def do_GET(self):
        path = self.path.split('?')[0]
        if path == '/products': self.send(200, PRODUCTS)
        elif path == '/health': self.send(200, {'status': 'ok'})
        else: self.send(404, {'error': 'not found'})
    def do_POST(self): self.send(405, {'error': 'method not allowed'})

if __name__ == '__main__':
    with ThreadingHTTPServer((os.getenv('BIND_HOST', '127.0.0.1'), int(os.getenv('PORT', '8081'))), Handler) as server:
        server.serve_forever()
